src/realm/sync/noinst/protocol_codec.hpp

```diff
@@ -524,60 +435,63 @@ class ServerProtocol {
     // The result of the parsing is handled by an object of type Connection.
     // Typically, Connection would be the Connection class from server.cpp
     template <class Connection>
-    void parse_message_received(Connection& connection, const char* data, std::size_t size)
+    void parse_message_received(Connection& connection, util::StringView msg_data)
     {
         util::Logger& logger = connection.logger;
 
-        util::MemoryInputStream in;
-        in.set_buffer(data, data + size);
-        in.unsetf(std::ios_base::skipws);
-        std::size_t header_size = 0;
-        std::string message_type;
-        in >> message_type;
+        auto report_error = [&](Error err, const auto fmt, auto&&... args) {
+            logger.error(fmt, args...);
+            connection.handle_protocol_error(err);
+        };
 
-        int protocol_version = connection.get_client_protocol_version();
-        static_cast<void>(protocol_version); // No divergent protocol behavior (yet).
+        auto parse_and_check_msg_data = [&](const auto what, char end_delim, auto&&... args) {
```

Comment:
maybe worth making this a utility function at file scope since it is used multiple times?

Reply:
hmm, well, it actually modifies msg_data, so it has state.